crates/common/rlp/decode.rs

```diff
@@ -534,6 +542,71 @@ pub fn static_left_pad<const N: usize>(data: &[u8]) -> Result<[u8; N], RLPDecode
     Ok(result)
 }
 
+/// Decodes a RLP integer limited by max_len into a u64.
+#[inline]
+fn decode_be_integer_u64(bytes: &[u8], max_len: usize) -> Result<u64, RLPDecodeError> {
+    debug_assert!(max_len <= 8);
+
+    if bytes.len() > max_len {
+        return Err(RLPDecodeError::InvalidLength);
+    }
+    if bytes.is_empty() {
+        return Ok(0);
+    }
+    if bytes[0] == 0 {
+        return Err(RLPDecodeError::MalformedData);
+    }
+
+    let mut value: u64 = 0;
+    for byte in bytes {
+        value = (value << 8) | (*byte as u64);
+    }
+    Ok(value)
+}
+
+/// Decodes a RLP integer into a u128, it's separate from the u64 to avoid paying the cost of using a u128.
+#[inline]
+fn decode_be_integer_u128(bytes: &[u8]) -> Result<u128, RLPDecodeError> {
+    if bytes.len() > 16 {
+        return Err(RLPDecodeError::InvalidLength);
+    }
+
+    if bytes.is_empty() {
+        return Ok(0);
+    }
+
+    if bytes[0] == 0 {
+        return Err(RLPDecodeError::MalformedData);
+    }
+
+    let mut value: u128 = 0;
+    for byte in bytes {
+        value = (value << 8) | (*byte as u128);
+    }
+    Ok(value)
+}
+
+/// Parses a RLP payload length
+#[inline]
+fn parse_be_length(bytes: &[u8]) -> Result<usize, RLPDecodeError> {
+    if bytes.is_empty() {
+        return Err(RLPDecodeError::InvalidLength);
+    }
+
+    if bytes[0] == 0 {
+        return Err(RLPDecodeError::MalformedData);
+    }
+
+    let mut value: usize = 0;
+    for byte in bytes {
+        value = value
+            .checked_mul(256)
+            .and_then(|v| v.checked_add(*byte as usize))
+            .ok_or(RLPDecodeError::InvalidLength)?;
+    }
```

Comment:
Probably just:
1. Check the length of the slice beforehand, it must not be greater than `size_of::<usize>()`;
2. Use shifts and binary ors.

In general, I wonder why this isn't just another call to `decode_be_integer_u64`.